### PR TITLE
fix: invalidate parent load functions

### PR DESCRIPTION
**Pull request by gterras**
Fixes: https://github.com/sveltejs/kit/issues/11696

also probably
https://github.com/sveltejs/kit/issues/11635
https://github.com/sveltejs/kit/issues/11663
https://github.com/sveltejs/kit/issues/11446
https://github.com/sveltejs/kit/issues/10209
https://github.com/sveltejs/kit/issues/10123
https://github.com/sveltejs/kit/issues/10457
https://github.com/sveltejs/kit/issues/10359
https://github.com/sveltejs/kit/issues/11745

See context: https://github.com/sveltejs/kit/pull/11268#issuecomment-1894232508

Most likely https://github.com/sveltejs/kit/pull/11268 introduced this behavior that is a blocker preventing an upgrade to SvelteKit2 for any project using "parent invalidation".

Draft status as it needs 2 tests: 
-  double invalidation (https://github.com/sveltejs/kit/issues/10359), is it technically possible to test this? 
-  parent invalidation synchronization   

I would like to dig into the tests but could someone confirm this PR is on the right tracks? 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [ x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

**Comment by kyngs** [comment]
I've tested this and can confirm that it fixes issue https://github.com/sveltejs/kit/issues/11696

File: packages/kit/src/runtime/client/client.js [hunk 1]
```diff
@@ -753,8 +752,10 @@ function has_changed(
 		if (params[param] !== current.params[param]) return true;
 	}
 
-	for (const href of uses.dependencies) {
-		if (invalidated.some((fn) => fn(new URL(href)))) return true;
+	if (!parent_changed) {
```
Comment (dummdidumm) [inline packages/kit/src/runtime/client/client.js:755]: Why would this logic only run if the parent hasn't changed? Wouldn't this result in false negatives?

File: packages/kit/src/runtime/client/client.js [hunk 2]
```diff
@@ -318,7 +318,6 @@ async function _invalidate() {
 		}
 	}
 
-	invalidated.length = 0;
```
Comment (dummdidumm) [inline packages/kit/src/runtime/client/client.js:321]: This would probably bring back #10677?

**Review by dummdidumm — COMMENTED** [review]
Thanks for making a start at this - though I'm not sure if this looks correct yet. In general the whole "when things are invalidated / reset" is hella hard to figure out. I have to think this through next week.

File: packages/kit/src/runtime/client/client.js (hunk 1 above)
Comment (gterras) [reply packages/kit/src/runtime/client/client.js:755]: What I understand of the whole thing is that every "parent routes" needs to go through the whole invalidation loop else their version of data is not updated. The logic in the file starts with the current page, then up a level until there is none. 

The fix in https://github.com/sveltejs/kit/pull/11268 would strictly break this loop after first iteration, so parents are never actually loaded. 

Having `!parent_changed` there is a mean  to ensure this is not the first iteration of the loop (meaning it's a parent), so it's data isn't actually invalidated (no need) but every routes still go through a invalidation process (so they get the updated data). 

So basically there is a `invalidation process` that do several things (invalidate the data and assign this data to the route being invalidated), and every segment of the page **need** to go through this process to receive the proper data, but only the current page segment have to actually invalidate data.

If you prevent parent segments from going into this loop they never get the updated data.

File: packages/kit/src/runtime/client/client.js (hunk 2 above)
Comment (gterras) [reply packages/kit/src/runtime/client/client.js:321]: I see tests failing in the PR but can't find a useful error

**Comment by gterras** [comment]
> Thanks for making a start at this - though I'm not sure if this looks correct yet. In general the whole "when things are invalidated / reset" is hella hard to figure out. I have to think this through next week.

Thanks for answering :pray:  Yes it really seems to be a tough topic and really to tackle without additional tests, I can try to add failing ones if you need it. This one especially 

> double invalidation (https://github.com/sveltejs/kit/issues/10359), is it technically possible to test this?

Starting from this one could be a softer approach since it's simpler but have similar roots.

File: packages/kit/src/runtime/client/client.js (hunk 2 above)
Comment (kyngs) [reply packages/kit/src/runtime/client/client.js:321]: It looks like #11268 did not fix that issue completely. Walking through the instructions in https://github.com/davideaster/sveltekit-invalidations I can still reproduce the error in the last step on the latest commit of sveltekit.

However, when I check out this PR, I can confirm that it brings back the first error from the reproduction repo I've linked above. The PR fixed that first error (`invalidate on a non-depends resource following invalidate with depends`).

**Comment by benmccann** [comment]
>  Yes it really seems to be a tough topic and really hard to tackle without additional tests, I can try to add failing ones if you need it.

That would be a huge help. I'm on vacation for the next week, so unfortunately not likely to make progress on this personally in the short-term. Tests would help expedite things so that some progress is made in the meantime
